include/wil/com.h

```diff
@@ -3325,6 +3325,80 @@ WI_NODISCARD auto make_range(IEnumXxx* enumPtr)
 #endif // WIL_HAS_CXX_17
 #endif // WIL_ENABLE_EXCEPTIONS
 
+#if WINAPI_FAMILY_PARTITION(WINAPI_PARTITION_DESKTOP | WINAPI_PARTITION_SYSTEM)
+#ifdef __WIL_WIN32_HELPERS_INCLUDED
+
+/** RAII support for making cross-apartment (or cross process) COM calls with a timeout applied to them.
+ * When this is active any timed out calls will fail with an RPC error code such as RPC_E_CALL_CANCELED.
+ * This is a shared timeout that applies to all calls made on the current thread for the lifetime of
+ * the wil::rpc_timeout object.
+~~~
+{
+  auto timeout = wil::rpc_timeout(5000);
+  remote_object->BlockingCOMCall();
+  remote_object->AnotherBlockingCOMCall();
+}
+~~~
+Include wil/win32_helpers.h before wil/com.h to use this.
+*/
+class rpc_timeout
+{
+public:
+    rpc_timeout(DWORD timeoutInMilliseconds):
+        m_threadId(GetCurrentThreadId())
+    {
+        m_cancelEnablementResult = CoEnableCallCancellation(nullptr);
+        if (SUCCEEDED(m_cancelEnablementResult))
+        {
+            m_timer.reset(CreateThreadpoolTimer(&rpc_timeout::timer_callback, this, nullptr));
+            LOG_LAST_ERROR_IF_NULL(m_timer.get());
+            if (m_timer)
+            {
+                FILETIME ft = filetime::get_system_time();
+                filetime::add(ft, filetime::convert_msec_to_100ns(timeoutInMilliseconds));
+                SetThreadpoolTimer(m_timer.get(), &ft, 0, 0);
+            }
+        }
+    }
+
+    ~rpc_timeout()
+    {
+        if (SUCCEEDED(m_cancelEnablementResult))
+        {
+            CoDisableCallCancellation(nullptr);
+        }
+
+        m_timer.reset();
+    }
+
+    bool timed_out() const
+    {
+        return m_timedOut;
+    }
+
+private:
+    // Disable use of new as this class should only be declared on the stack, never the heap.
+    void *operator new(size_t) = delete;
+    void *operator new[](size_t) = delete;
+
+    static void timer_callback(PTP_CALLBACK_INSTANCE /*instance*/, PVOID context, PTP_TIMER /*timer*/)
+    {
+        // The timer is waited upon during destruction so it is safe to rely on the this pointer in context.
+        rpc_timeout* self = static_cast<rpc_timeout*>(context);
+        if (SUCCEEDED(CoCancelCall(self->m_threadId, 0)))
+        {
+            self->m_timedOut = true;
+        }
+    }
+
+    wil::unique_threadpool_timer_nocancel m_timer;

```

Comment:
Is the `nocancel` version the one you want? What happens if the timer fires right after the COM call returns to you but before the dtor finishes here?

Reply:
The destructor would block until the callback is completed, preventing the `this` pointer from being used after destruction.  The callback doesn't do significant work so it shouldn't block for long.

Reply:
One thing to keep in mind - although I don't necessarily think it's an issue here - is that members construct in-order before the constructor runs and destruct in reverse order after the destructor runs. Effectively, this means that the callback could be running while the rest of this object has been destructed (since this is the first member). You're not doing anything crazy like setting `m_threadId` to zero and writing to `m_timedOut` should be no-op, however this also means that it's possible for `CoDisableCallCancellation` to be called before `CoCancelCall` is called in the callback. I'd hope that COM handles this gracefully and no-ops/returns failure for `CoCancelCall` and so no harm is done; just something to consider.

Reply:
That's a great point.  Reordering fields should eliminate that risk.  I'll do that and add a comment describing it.